### PR TITLE
transposed convolution implementation

### PR DESCRIPTION
**Pull request by levskaya**
This implements the upsampling 'transposed convolutions' in N-dimensions for any strides and for same, valid, or manually specified padding.  It was written to mimic the behavior of keras.layers.Conv2dTranspose and other similar grad-of-conv based transposed-convolutions as most people are familiar with these.  However, this implementation is __direct__, _not_ implemented in terms of a vjp of conv.

In fact, the default mode of operation is to calculate a 'fractionally-strided' conv w. LHS-dilation and __not__ to bother flipping and swapping the kernel (rhs) to mimic a true transpose, as this is just wasteful computationally, confuses input/output channels, and has no representational advantage for any use-case in neural nets.  True transposed-convs _can_ be calculated by passing `tranpose_kernel=True` to perfectly mimic the keras Conv2dTranspose applied to the given kernel if for some obscure reason you wanted to do that.  (e.g. I unit test against a vjp-of-conv.)

I retained the conv-transpose name only because most people call them that, but we could change that.

### CODE REVIEW

**Comment by levskaya** [comment]
weird looks like two of travis build/test jobs just randomly timed out...

**Review by mattjj — APPROVED** [review]
This is fantastic, thanks so much for doing it. The code is clean and the tests are thorough! Great docstring too; it’s really helpful to relate to a standard interface like Keras.

**Comment by mattjj** [comment]
Travis has done that to me too. Seems random.
